query04.sql

```diff
@@ -0,0 +1,27 @@
+/*
+Using the shapes.txt file from GTFS bus feed, find the two routes with the longest trips. In the final query, give the trip_headsign that corresponds to the shape_id of this route and the length of the trip.
+*/
+
+
+/*
+Explanation:
+
+My original approach, which I believe to be the correct approach, was to measure the distance between the first and last stops for each route by using ST_Distance between the MAX shape_pt_sequence and MIN shape_pt_sequence for each shape_id. After several attempts, however, I could not determine how to isolate the rows containing the MAX and MIN shape_pt_sequence for each shape_id.
```

Comment:
Ah, makes sense. The approach I was looking for was to find the total route length, either by using `ST_MakeLine`, or by summing all the distances between subsequent stops:

_ST_MakeLine_:
```sql
WITH septa_bus_lines AS (
    SELECT shape_id,
           ST_MakeLine(the_geom ORDER BY shape_pt_sequence) AS shape_line
    FROM septa_bus_shapes
    GROUP BY shape_id
    
)

SELECT DISTINCT
    trip_headsign,
    ST_Length(ST_Transform(shape_line, 32129)) AS trip_length
FROM septa_bus_lines
JOIN septa_bus_trips USING (shape_id)
ORDER BY trip_length DESC
LIMIT 2
```

_SUM_:
```sql
WITH septa_bus_shape_segments AS (
    SELECT a.shape_id,
           ST_Transform(a.the_geom, 32129) AS geom_1,
           ST_Transform(b.the_geom, 32129) AS geom_2
    FROM septa_bus_shapes AS a
    JOIN septa_bus_shapes AS b USING (shape_id)
    WHERE a.shape_pt_sequence = b.shape_pt_sequence - 1
),

septa_bus_shape_lengths AS (
    SELECT shape_id,
           SUM(ST_Distance(geom_1, geom_2)) AS trip_length
    FROM septa_bus_shape_segments
    GROUP BY shape_id
)

SELECT DISTINCT
    trip_headsign,
    trip_length
FROM septa_bus_shape_lengths
JOIN septa_bus_trips USING (shape_id)
ORDER BY trip_length DESC
LIMIT 2
```